src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+                // TODO: Should error abort if unsupported constraint detected.
+                // Currently warning is a workaround for making ontology
+                // working, as in some cases existential constraints generated.
+                // Should investigate on this, and change this to ErrorAbort
+                // when eliminated unsupported constraints.
+                logger.fine(
+                        "Unsupported constraint detected! Constraint type: "
+                                + constraint.getClass().getSimpleName());
+                continue;
+            }
+
+            Expr simplifiedConstraint = serializedConstraint.simplify();
+
+            if (simplifiedConstraint.isTrue()) {
+                // This only works if the BoolExpr is directly the value Z3True.
+                // Still a good filter, but doesn't filter enough.
+                // EG: (and (= false false) (= false false) (= 0 0) (= 0 0) (= 0 0))
+                // Skip tautology.
+                current++;
+                continue;
+            }
+
+            if (simplifiedConstraint.isFalse()) {
+                final ToStringSerializer toStringSerializer = new ToStringSerializer(false);
+                throw new BugInCF(
+                        "impossible constraint: "
+                                + constraint.serialize(toStringSerializer)
+                                + "\nSerialized:\n"
+                                + serializedConstraint);
+            }
+
+            String clause = simplifiedConstraint.toString();
+
+            if (!optimizingMode && getUnsatCore) {
+                // add assertions with names, for unsat core dump
+                String constraintName = constraint.getClass().getSimpleName() + current;
+
+                constraintSmtFileContents.append("(assert (! ");
+                constraintSmtFileContents.append(clause);
+                constraintSmtFileContents.append(" :named " + constraintName + "))\n");
+
+                // add constraint to serialized constraints map, so that we can
+                // retrieve later using the constraint name when outputting the unsat core
+                serializedConstraints.put(constraintName, constraint);
+            } else {
+                constraintSmtFileContents.append("(assert ");
+                constraintSmtFileContents.append(clause);
+                constraintSmtFileContents.append(")\n");
+            }
+
+            current++;
+        }
+
+        String constraintSmt = constraintSmtFileContents.toString();
+
+        smtFileContents.append(constraintSmt);
+    }
+
+    protected void encodeAllSoftConstraints() {
+    	final Z3SmtSoftConstraintEncoder<SlotEncodingT, SlotSolutionT> encoder = formatTranslator.createSoftConstraintEncoder();
+        smtFileContents.append(encoder.encodeAndGetSoftConstraints(constraints));
+    }
+
+    protected void encodeSlotPreferenceConstraint(Slot varSlot) {
```

Comment:
VariableSlot as parameter.
Why is this a separate method? Is this expected to be overridden?